File: TOH-Base/src/main/java/com/tohsoft/base/mvp/utils/language/LocaleManager.java
package com.tohsoft.base.mvp.utils.language;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.os.LocaleList;

import com.utility.DebugLog;
import com.utility.SharedPreference;

import java.util.Locale;

public class LocaleManager {
    public static String LANGUAGE_SELECTED = "LANGUAGE_SELECTED";
    public static String MODE_AUTO = "auto";

    public static Context setLocale(Context c) {
        return updateResources(c, getLanguage(c));
    }

    public static Context setNewLocale(Context c, String language) {
        persistLanguage(c, language);
        return updateResources(c, language);
    }

    public static String getLanguage(Context c) {
        return SharedPreference.getString(c, LANGUAGE_SELECTED, MODE_AUTO);
    }

    @SuppressLint("ApplySharedPref")
    private static void persistLanguage(Context c, String language) {
        SharedPreference.setString(c, LANGUAGE_SELECTED, language);
    }

    private static Context updateResources(Context context, String language) {
        Locale locale;
        if (language.equals(MODE_AUTO)) {
            locale = Resources.getSystem().getConfiguration().locale;
        } else {
            if (language.equals("zh-rCN")||language.equals("zh")) {
                locale = Locale.SIMPLIFIED_CHINESE;
            } else if (language.equals("zh-rTW")) {
                locale = Locale.TRADITIONAL_CHINESE;
            } else {
                String[] spk = language.split("-");
                if (spk.length > 1) {
                    locale = new Locale(spk[0], spk[1]);
                } else {
                    locale = new Locale(spk[0]);
                }
            }
        }
        Locale.setDefault(locale);
        DebugLog.logd("Change language:\nlanguage: " + language + "\nlocale: " + locale.getLanguage());

        return updateResourcesLocaleLegacy(context, locale);
    }

    private static Context updateResourcesLocaleLegacy(Context context, Locale locale) {
        Resources resources = context.getResources();
        Configuration configuration = resources.getConfiguration();
        configuration.setLocale(locale);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            LocaleList localeList = new LocaleList(locale);
            LocaleList.setDefault(localeList);
        }
        resources.updateConfiguration(configuration, resources.getDisplayMetrics());
        return context;
    }

    public static Locale getLocale(Resources res) {
        Configuration config = res.getConfiguration();
        return Build.VERSION.SDK_INT >= 24 ? config.getLocales().get(0) : config.locale;
    }
}

File: TOH-Base/src/main/java/com/tohsoft/base/mvp/utils/language/LocaleManager.kt
package com.tohsoft.base.mvp.utils.language

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Resources
import android.os.Build
import android.os.LocaleList
import com.utility.SharedPreference
import java.util.*

object LocaleManager {
    var LANGUAGE_SELECTED = "LANGUAGE_SELECTED"
    @JvmField
    var MODE_AUTO = "auto"
    fun setLocale(c: Context): Context {
        return updateResources(c, getLanguage(c))
    }

    @JvmStatic
    fun setNewLocale(c: Context, language: String): Context {
        persistLanguage(c, language)
        return updateResources(c, language)
    }

    @JvmStatic
    fun getLanguage(c: Context?): String {
        return SharedPreference.getString(c, LANGUAGE_SELECTED, MODE_AUTO)
    }

    @SuppressLint("ApplySharedPref")
    private fun persistLanguage(c: Context, language: String) {
        SharedPreference.setString(c, LANGUAGE_SELECTED, language)
    }

    private fun updateResources(context: Context, language: String): Context {
        val locale: Locale = if (language == MODE_AUTO) {
            Resources.getSystem().configuration.locale
        } else {
            if (language == "zh-rCN" || language == "zh") {
                Locale.SIMPLIFIED_CHINESE
            } else if (language == "zh-rTW") {
                Locale.TRADITIONAL_CHINESE
            } else {
                val spk = language.split("-".toRegex()).toTypedArray()
                if (spk.size > 1) {
                    Locale(spk[0], spk[1])
                } else {
                    Locale(spk[0])
                }
            }
        }
        Locale.setDefault(locale)
        return updateResourcesLocaleLegacy(context, locale)
    }

    private fun updateResourcesLocaleLegacy(context: Context, locale: Locale): Context {
        val resources = context.resources
        val configuration = resources.configuration
        configuration.setLocale(locale)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val localeList = LocaleList(locale)
            LocaleList.setDefault(localeList)
        }
        resources.updateConfiguration(configuration, resources.displayMetrics)
        return context
    }

    @JvmStatic
    fun getLocale(res: Resources): Locale {
        val config = res.configuration
        return if (Build.VERSION.SDK_INT >= 24) config.locales[0] else config.locale
    }
}